### ISSUE TITLE
SingleGlass page type

### ISSUE THREAD

**Issue by jdaw1** [jdaw1/placemat#162]
Nice to post online might be a GIF animation of all the glass circles (second image). 

<div align="center">

![20220705_SingleGlass](https://user-images.githubusercontent.com/49350648/177855739-ec9ea267-becf-4078-8e22-c3034bd0a40e.png)&nbsp;&nbsp;![20220705_SingleGlass](https://user-images.githubusercontent.com/49350648/177428151-9e96779b-bc75-4f73-9783-51afdcee0807.gif)

</div>

The natural way to proceed is to have the software make single-glass pages. For each page, width&nbsp;= height&nbsp;= 2&nbsp;&times;&nbsp;radius, with the circle centred. GIFing is then easy, such as via [GraphicConverter](https://www.lemkesoft.de/en/products/graphicconverter/)’s *File* > *New* > _Animation using all Pages of the multipage Document (TIFF,PDF)_.

Currently the PostScript part can be done by using `PrologueCode` to inject code (software version &ge;&nbsp;04&nbsp;July 2022). For a few goes this will be the method, to allow learning. If/when in the software, some of the functionality would move into the existing `ShowPage` routine, and the `mark … /OUT pdfmark` would also be separated.

Questions:
* Should the `TypeOfPagesBeingRendered` be named &lsquo;SingleGlass&rsquo;?
* Will there be a need for multiple new parameters, or would it be acceptable to re-use `PrePourShowBackgroundTexts` and `PrePourShowDroplets`?
* Do people agree that it should be controlled by a Boolean, rather than a `…NumCopies` integer?
* There is a Boolean `BottleWrapSuppressOtherPageTypes`. This page type needs similar. But the two would interact non-intuitively. What should be done?

```PostScript
% From https://github.com/jdaw1/placemat/issues/162
/PrologueCode
{
	/TestingSuppressPageTypes [ /Glasses 0  /TastingNotes 0  /VoteRecorder 0  /DecantingNotes 0  /Accounts 0  /CorkDisplay 0  /NeckTags 0  /PrePour 0  /PlaceName 0  /DecanterLabels 0  /BottleWrap 0  /StickyLabels 0  /DistillerLog ] store
	0 SheetLengths {add} forall  dup 0 gt
	{
		17 dict begin
		mark   /Count 3 -1 roll    /Action << >>    /F 0    /Title (Single-circle extracts)    /OUT pdfmark
		0  1  NumSheets 1 sub
		{
			/SheetNum exch def
			//false //false //false //false PaperType Orientation SetPaperSize   //true DefStoreMgns
			0  1  SheetLengths SheetNum get 1 sub
			{
				WithinPage-WithinTitles-def
				CircleNonEmpty SheetNum get WithinPage get
				{
					<< /PageSize [Radii SheetNum get 2 mul  dup] >> setpagedevice
					Radii SheetNum get  dup  translate  newpath
					/MatrixGlasses matrix currentmatrix def
					0  0  Radii SheetNum get  0  360  arc  1 setgray fill

					PrePourShowBackgroundTexts   Droplets {PrePourShowDroplets or} if
					{
						ClipSave  newpath  0  0  Radii SheetNum get  0  360  arc  clip  newpath
						/TypeOfPagesBeingRendered /Glasses store
						GlassPositions SheetNum get WithinPage get {neg} forall translate
						/MatrixGlassesPage matrix currentmatrix def
						BackgroundTextsGlasses PrePourShowBackgroundTexts and {BackgroundTextsGlassesForms SheetNum get execform} if
						Droplets PrePourShowDroplets and
						{
							SheetNum
							GSave  newpath
							MatrixGlasses setmatrix
							Radii SheetNum get dup   dup dup neg  dup dup  dup 4 index  moveto lineto lineto lineto
							MatrixGlassesPage setmatrix   pathbbox	GRestore
							WithinPage  0.0  DropletsPaint
						} if  % Droplets PrePourShowDroplets and
						ClipRestore
					} if  % PrePourShowBackgroundTexts ...PrePourShowDroplets or
					MatrixGlasses setmatrix

					/TypeOfPagesBeingRendered /Glasses store
					OutlineTitles
					{
						OutlineForms SheetNum get WithinPage get   execform
					} if  % OutlineTitles ...

					Spirals {SpiralForms SheetNum get WithinPage get execform} if
					CrossHatchingInside {CrossHatchingInsideForms SheetNum get WithinPage get  execform} if

					/PaintBackgroundInsideGlassCircles load dup length 0 gt
					{
						/TypeOfPagesBeingRendered /TastingNotes store
						GSave newpath  0 0 Radii SheetNum get 0 360 arc  clip newpath
						execU   GRestore /TypeOfPagesBeingRendered /Glasses store
					} {pop} ifelse  % PaintBackgroundInsideGlassCircles ... length 0 gt
					TitleAboveBelowOverForms SheetNum get WithinPage get  execform  AnnotateGlass
					CirclearrayForms SheetNum get WithinPage get   execform

					/PageDestName  mark (SingleCircle_) SheetNum (_) WithinPage ConcatenateToMark cvn  def
					mark /Dest PageDestName  /View [/Fit]   /DEST  pdfmark
					mark
						/Action /GoTo    /Dest PageDestName    /F 0
						/Title mark ()
							[ Titles  Belowtitles  Abovetitles  Overtitles  FillTitles {FillTexts} if ]
								{WithinTitles get PDFDocEncodingify TrimSpaces dup length 0 gt {(, )} {pop} ifelse}
							forall pop ConcatenateToMark TruncateTo255
							dup length 0 eq {pop mark WithinTitles (=WithinTitles) ConcatenateToMark} if
					/OUT pdfmark
					showpage
				} if  % CircleNonEmpty ...
			} for  % WithinPage, WithinTitles
		} bind for  % SheetNum
		end
	} {pop} ifelse  % any glasses
} bind def  % /PrologueCode, which executed once, just before painting pages
```

**Comment by jdaw1** [comment]
Code in first post updated, mostly so that its call to `DropletsPaint` includes the new _LengthsDistance_ parameter, but also with other tidyings.

**Comment by jdaw1** [comment]
Closed by commit b86f29bac90d5d88ab3df9bb443d6eb3506792ee

The code at the top had an error: the `CircleNonEmpty` test worked for pages, but didn’t the first `mark /Count … /OUT pdfmark` assumed all were non-empty. Happily, `ShowPage` effectively handles this.
